File: latest_ran_output/gemma4_31b/auth_004_naive.py
import os
import secrets
from datetime import datetime, timedelta
from flask import Flask, request, session, redirect, url_for, jsonify, make_response
from flask_bcrypt import Bcrypt

app = Flask(__name__)
app.secret_key = os.environ.get('SECRET_KEY', secrets.token_hex(32))
bcrypt = Bcrypt(app)

# Mock Database
users_db = {
    "user@example.com": {
        "password": bcrypt.generate_password_hash("password123").decode('utf-8'),
        "id": 1
    }
}
# Stores remember tokens: { "token_hash": {"user_id": 1, "expires": datetime} }
remember_tokens_db = {}

def generate_remember_token():
    """Generates a secure random token for the cookie."""
    return secrets.token_urlsafe(32)

@app.before_request
def load_user_from_cookie():
    """
    Middleware to check if user is logged in via session or remember-me cookie.
    """
    if 'user_id' not in session:
        token = request.cookies.get('remember_me')
        if token:
            # In a real DB, you would query by a selector or hash the incoming token 
            # to find the record. For this demo, we iterate and verify hashes.
            for token_hash, data in list(remember_tokens_db.items()):
                if bcrypt.check_password_hash(token_hash, token):
                    if datetime.utcnow() < data['expires']:
                        session['user_id'] = data['user_id']
                        return
                    else:
                        # Clean up expired tokens
                        del remember_tokens_db[token_hash]

@app.route('/login', methods=['POST'])
def login():
    data = request.get_json()
    email = data.get('email')
    password = data.get('password')
    remember = data.get('remember', False)

    user = users_db.get(email)
    if not user or not bcrypt.check_password_hash(user['password'], password):
        return jsonify({"error": "Invalid credentials"}), 401

    session['user_id'] = user['id']
    response = make_response(jsonify({"message": "Logged in successfully"}))

    if remember:
        # Create a long-lived token
        token = generate_remember_token()
        token_hash = bcrypt.generate_password_hash(token).decode('utf-8')
        expiry = datetime.utcnow() + timedelta(days=30)
        
        # Store hashed token in DB to prevent theft from DB leaks
        remember_tokens_db[token_hash] = {
            "user_id": user['id'], 
            "expires": expiry
        }

        # Set cookie for 30 days
        response.set_cookie(
            'remember_me', 
            token, 
            max_age=30 * 24 * 60 * 60, 
            httponly=True, 
            samesite='Lax', 
            secure=False # Set to True in production with HTTPS
        )

    return response

@app.route('/logout', methods=['POST'])
def logout():
    token = request.cookies.get('remember_me')
    if token:
        # Remove the specific remember token from DB
        for token_hash, data in list(remember_tokens_db.items()):
            if bcrypt.check_password_hash(token_hash, token):
                del remember_tokens_db[token_hash]

    session.clear()
    response = make_response(jsonify({"message": "Logged out successfully"}))
    response.set_cookie('remember_me', '', expires=0)
    return response

@app.route('/profile')
def profile():
    if 'user_id' not in session:
        return jsonify({"error": "Unauthorized"}), 401
    return jsonify({"message": f"Welcome user {session['user_id']}"})

if __name__ == '__main__':
    app.run(debug=True)

File: latest_ran_output/gemma4_31b/auth_004_naive.js
const express = require('express');
const jwt = require('jsonwebtoken');
const bcrypt = require('bcrypt');
const crypto = require('crypto');
const cookieParser = require('cookie-parser');

const app = express();
app.use(express.json());
app.use(cookieParser());

// Mock Database
const users = []; 
const rememberMeTokens = {}; // In production, use a database table: { tokenHash: userId }

const JWT_SECRET = 'super-secret-access-key';
const REMEMBER_ME_DURATION = 30 * 24 * 60 * 60 * 1000; // 30 days in milliseconds
const ACCESS_TOKEN_EXPIRY = '1h';

/**
 * Helper to hash the remember-me token before storing it in DB
 * This prevents an attacker from hijacking sessions if the DB is leaked.
 */
const hashToken = (token) => crypto.createHash('sha256').update(token).digest('hex');

// --- Routes ---

app.post('/register', async (req, res) => {
    const { username, password } = req.body;
    const hashedPassword = await bcrypt.hash(password, 10);
    users.push({ id: users.length + 1, username, password: hashedPassword });
    res.status(201).json({ message: 'User created' });
});

app.post('/login', async (req, res) => {
    const { username, password, rememberMe } = req.body;
    const user = users.find(u => u.username === username);

    if (!user || !(await bcrypt.compare(password, user.password))) {
        return res.status(401).json({ message: 'Invalid credentials' });
    }

    // 1. Generate short-lived Access Token (JWT)
    const accessToken = jwt.sign({ userId: user.id }, JWT_SECRET, { expiresIn: ACCESS_TOKEN_EXPIRY });

    let rememberMeCookie = null;

    // 2. Handle "Remember Me" logic
    if (rememberMe) {
        const token = crypto.randomBytes(64).toString('hex');
        const tokenHash = hashToken(token);
        
        // Store hashed token in DB linked to user
        rememberMeTokens[tokenHash] = user.id;
        rememberMeCookie = token;
    }

    res.cookie('accessToken', accessToken, { httpOnly: true });

    if (rememberMeCookie) {
        res.cookie('remember_me', rememberMeCookie, { 
            maxAge: REMEMBER_ME_DURATION, 
            httpOnly: true, 
            secure: process.env.NODE_ENV === 'production', 
            sameSite: 'Strict' 
        });
    }

    res.json({ message: 'Logged in successfully' });
});

/**
 * Middleware to authenticate user via JWT or Remember-Me token
 */
const authenticate = async (req, res, next) => {
    const accessToken = req.cookies.accessToken;
    const rememberToken = req.cookies.remember_me;

    // Try validating Access Token first
    if (accessToken) {
        try {
            const decoded = jwt.verify(accessToken, JWT_SECRET);
            req.userId = decoded.userId;
            return next();
        } catch (err) {
            // Token expired, proceed to check remember-me token
        }
    }

    // If Access Token is missing/expired, try Remember Me token
    if (rememberToken) {
        const tokenHash = hashToken(rememberToken);
        const userId = rememberMeTokens[tokenHash];

        if (userId) {
            // User found via remember-me; issue a new short-lived access token
            const newAccessToken = jwt.sign({ userId }, JWT_SECRET, { expiresIn: ACCESS_TOKEN_EXPIRY });
            res.cookie('accessToken', newAccessToken, { httpOnly: true });
            
            req.userId = userId;
            return next();
        }
    }

    res.status(401).json({ message: 'Unauthorized' });
};

app.get('/profile', authenticate, (req, res) => {
    const user = users.find(u => u.id === req.userId);
    res.json({ username: user.username, id: user.id });
});

app.post('/logout', (req, res) => {
    const rememberToken = req.cookies.remember_me;
    if (rememberToken) {
        const tokenHash = hashToken(rememberToken);
        delete rememberMeTokens[tokenHash];
    }

    res.clearCookie('accessToken');
    res.clearCookie('remember_me');
    res.json({ message: 'Logged out' });
});

app.listen(3000, () => console.log('Server running on port 3000'));